src/components/CheckBoxFilterContainer.ts

```diff
@@ -0,0 +1,203 @@
+            this.renderComponent()
+        );
+    }
+
+    componentWillUpdate(nextProps: ContainerProps, nextState: ContainerState) {
+        console.log("willUpdate"); // tslint:disable-line
+        this.alertMessage = this.validate(nextProps, nextState.targetListView);
+        this.applyFilter(nextProps.defaultChecked);
+    }
+
+    componentDidUpdate(_prevProps: ContainerProps, prevState: ContainerState) {
+        if (this.state.listViewAvailable && !prevState.listViewAvailable) {
+            this.applyFilter(this.props.defaultChecked);
+        }
+    }
+
+    private renderAlert() {
+        return createElement(Alert, {
+            className: "widget-checkbox-filter-alert",
+            message: this.alertMessage
+        });
+    }
+
+    private renderComponent(): ReactElement<CheckboxFilterProps> {
+        if (!this.alertMessage) {
+            return createElement(CheckboxFilter, {
+                handleChange: this.applyFilter,
+                isChecked: this.props.defaultChecked
+            });
+        }
+
+        return null;
+    }
+
+    private applyFilter(isChecked: boolean) {
+        // construct constraint based on checked
+        const constraint = this.getConstraint(isChecked);
+        if (this.dataSourceHelper)
+            this.dataSourceHelper.setConstraint(this.props.friendlyId, constraint);
+    }
+
+    private getConstraint(isChecked: boolean) {
+        const { targetListView } = this.state;
+        if (targetListView && targetListView._datasource) {
+            const attribute = isChecked ? this.props.attribute : this.props.unCheckedAttribute;
+            const filterBy = isChecked ? this.props.filterBy : this.props.unCheckedFilterBy;
+            const constraint = isChecked ? this.props.constraint : this.props.unCheckedConstraint;
+            const attributeValue = isChecked ? this.props.attributeValue : this.props.unCheckedAttributeValue;
+
+            if (filterBy === "XPath") {
+                return constraint.indexOf(`[%CurrentObject%]'`) !== -1
+                    ? constraint.replace(`'[%CurrentObject%]'`, this.props.mxObject.getGuid())
+                    : constraint;
+            } else if (filterBy === "attribute") {
+                return this.getAttributeConstaint(attribute, attributeValue);
+            } else {
+               return "";
+            }
+        }
+    }
+
+    private getAttributeConstaint(attribute: string, attributeValue: string): string | HybridConstraint {
+        const { targetListView } = this.state;
+
+        if (window.device) {
+            const constraints: HybridConstraint = {
+                attribute,
+                operator: "contains",
+                path: this.props.listViewEntity,
+                value: attributeValue
+            };
+            return constraints;
+        }
+        if (targetListView && targetListView._datasource) {
+            const mxObject = targetListView._datasource._pageObjs[0];
```

Comment:
Dont take something from the API if it is available in the public API
https://apidocs.mendix.com/7/client/mx.meta.html

``` ts
const entityMeta = mx.meta.getEntity(this.listViewEntity);
entityMeta.isEnum(attribute); // ETC
```

Reply:
Done